### PR TITLE
make glooctl check work for kube gateways

### PR DESCRIPTION
**Pull request by jenshu**
# Description

Check Kubernetes Gateway API resources during `glooctl check`, as long as we detect that the kube gateway integration is enabled in the given gloo installation. We assume kube gateway integration is enabled if 1) we find Gateway CRDs on the cluster and 2) a default GatewayParameters exists in the install namesapce.

Other cleanup:
- Standardized the casing of the `glooctl check` output and use the actual CRD names, e.g. `Checking auth configs... OK` is now `Checking AuthConfigs... OK`
- Removed unused KUBE2E_TESTS var from setup-kind. As of https://github.com/solo-io/gloo/pull/9636 it's not used anymore in the script.

Output :
```
Checking Deployments... OK
Checking Pods... OK
Checking Upstreams... OK
Checking UpstreamGroups... OK
Checking AuthConfigs... OK
Checking RateLimitConfigs... OK
Checking VirtualHostOptions... OK
Checking RouteOptions... OK
Checking Secrets... OK
Checking VirtualServices... OK
Checking Gateways... OK
Checking Proxies... OK

Detected Kubernetes Gateway integration!
Checking kube GatewayClasses... OK
Checking kube Gateways... OK
Checking kube HTTPRoutes... OK

Skipping Gloo Instance check -- Gloo Federation not detected.
No problems detected.
```
```
Checking Deployments... OK
Checking Pods... OK
Checking Upstreams... OK
Checking UpstreamGroups... OK
Checking AuthConfigs... OK
Checking RateLimitConfigs... OK
Checking VirtualHostOptions... OK
Checking RouteOptions... OK
Checking Secrets... OK
Checking VirtualServices... OK
Checking Gateways... OK
Checking Proxies... OK

Skipping Gloo Instance check -- Gloo Federation not detected.
Error: 1 error occurred:
	* unable to determine if kube gateway is enabled: KubeGateway is enabled but the Kubernetes Gateway CRDs are not applied. Please fix this by running `kubectl apply -f https://github.com/kubernetes-sigs/gateway-api/releases/download/v1.0.0/standard-install.yaml`
```

# Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: .github/workflows/composite-actions/regression-tests/action.yaml [hunk 1]
```diff
@@ -22,7 +22,6 @@ runs:
   - name: Setup test env
     shell: bash
     env:
-      KUBE2E_TESTS: ${{ matrix.kube-e2e-test-type }}
```
Comment (jenshu) [inline .github/workflows/composite-actions/regression-tests/action.yaml:25]: not used anymore by setup-kind

File: docs/content/reference/cli/glooctl_check.md [hunk 2]
```diff
@@ -17,7 +17,7 @@ glooctl check [flags]
 ### Options
 
 ```
-  -x, --exclude strings                   check to exclude: (deployments, pods, upstreams, upstreamgroup, auth-configs, rate-limit-configs, secrets, virtual-services, gateways, proxies, xds-metrics)
+  -x, --exclude strings                   check to exclude: (deployments, pods, upstreams, upstreamgroup, auth-configs, rate-limit-configs, virtual-host-options, route-options, secrets, virtual-services, gateways, proxies, xds-metrics)
```
Comment (jenshu) [inline docs/content/reference/cli/glooctl_check.md:20]: we already support excluding these, but they were missing from the description here

File: projects/gloo/cli/pkg/cmd/check/gloo_instances.go [hunk 3]
```diff
@@ -60,16 +61,19 @@ func CheckMulticlusterResources(ctx context.Context, printer printers.P, opts *o
...
+		printGlooInstanceCheckSummary("Pods", glooInstance.Spec.GetCheck().GetPods())
+		printGlooInstanceCheckSummary("Settings", glooInstance.Spec.GetCheck().GetSettings())
+		printGlooInstanceCheckSummary("Upstreams", glooInstance.Spec.GetCheck().GetUpstreams())
+		printGlooInstanceCheckSummary("UpstreamGroups", glooInstance.Spec.GetCheck().GetUpstreamGroups())
+		printGlooInstanceCheckSummary("AuthConfigs", glooInstance.Spec.GetCheck().GetAuthConfigs())
+		printGlooInstanceCheckSummary("RateLimitConfigs", glooInstance.Spec.GetCheck().GetRateLimitConfigs())
+		printGlooInstanceCheckSummary("VirtualServices", glooInstance.Spec.GetCheck().GetVirtualServices())
+		printGlooInstanceCheckSummary("RouteRables", glooInstance.Spec.GetCheck().GetRouteTables())
+		printGlooInstanceCheckSummary("Gateways", glooInstance.Spec.GetCheck().GetGateways())
+		printGlooInstanceCheckSummary("MatchableHttpGateways", glooInstance.Spec.GetCheck().GetMatchableHttpGateways())
```
Comment (jenshu) [inline projects/gloo/cli/pkg/cmd/check/gloo_instances.go:74]: there were some fields on `glooInstance.Spec.GetCheck()` that we weren't previously displaying (RLC, Matchable*Gateways) so added them here

File: projects/gloo/cli/pkg/cmd/check/gloo_instances.go [hunk 4]
```diff
@@ -101,12 +105,8 @@ type unstructuredGlooInstanceReader struct {
 }
 
 func getUnstructuredGlooInstanceReader(cfg *rest.Config) (*unstructuredGlooInstanceReader, error) {
-	scheme := scheme.Scheme
```
Comment (jenshu) [inline projects/gloo/cli/pkg/cmd/check/gloo_instances.go:104]: moved to `init()` to avoid race (since we need to call multiple AddToSchemes now)

File: projects/gloo/cli/pkg/cmd/check/internal/checks.go [hunk 5]
```diff
@@ -6,120 +6,25 @@ import (
...
 	"k8s.io/apimachinery/pkg/api/meta"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"sigs.k8s.io/controller-runtime/pkg/client/config"
 	gwv1 "sigs.k8s.io/gateway-api/apis/v1"
 	gwclient "sigs.k8s.io/gateway-api/pkg/client/clientset/versioned"
 )
 
-// Checks whether the cluster that the kubeconfig points at is available
-// The timeout for the kubernetes client is set to a low value to notify the user of the failure
-func CheckConnection(ctx context.Context, _ printers.P, opts *options.Options) error {
```
Comment (jenshu) [inline projects/gloo/cli/pkg/cmd/check/internal/checks.go:26]: CheckConnection and CheckDeployment are already done as part of the core `glooctl check` so removed them from the kubegateway checks

File: projects/gloo/cli/pkg/cmd/check/root.go [hunk 6]
```diff
@@ -73,30 +81,37 @@ func RootCmd(opts *options.Options, optionsFunc ...cliutils.OptionsFunc) *cobra.
...
+			}
 
+			// check kube gateway resources
+			err = CheckKubeGatewayResources(ctx, printer, opts)
 			if err != nil {
-				// Not returning error here because this shouldn't propagate as a standard CLI error, which prints usage.
+				multiErr = multierror.Append(multiErr, err)
+			}
+
+			// check gloo fed resources (TODO this should return errors too)
```
Comment (jenshu) [inline projects/gloo/cli/pkg/cmd/check/root.go:98]: wanted to make this similar to CheckResources and CheckKubeGatewayResources above and return errors (instead of just printing errors) but don't plan to handle this refactor in this PR

File: projects/gloo/cli/pkg/cmd/check/root.go [hunk 7]
```diff
@@ -73,30 +81,37 @@ func RootCmd(opts *options.Options, optionsFunc ...cliutils.OptionsFunc) *cobra.
...
+			CheckMulticlusterResources(ctx, printer, opts)
+
+			if multiErr.ErrorOrNil() != nil {
+				for _, err := range multiErr.Errors {
+					printer.AppendError(fmt.Sprint(err))
```
Comment (jenshu) [inline projects/gloo/cli/pkg/cmd/check/root.go:103]: moved the AppendErrors here to ensure the aggregated errors returned from the Check* funcs all get printed to json output

File: projects/gloo/cli/pkg/cmd/check/root.go [hunk 8]
```diff
@@ -73,30 +81,37 @@ func RootCmd(opts *options.Options, optionsFunc ...cliutils.OptionsFunc) *cobra.
...
+				}
+
+				// when output type is table, the returned errors get printed directly as cli output.
 				if opts.Top.Output.IsTable() {
-					return err
+					return multiErr.ErrorOrNil()
```
Comment (jenshu) [inline projects/gloo/cli/pkg/cmd/check/root.go:108]: make sure all errors get printed in table format

File: projects/gloo/cli/pkg/cmd/check/kube_gateway.go [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+			multiErr = multierror.Append(multiErr, err)
+		}
+	}
+
+	return multiErr.ErrorOrNil()
+}
+
+// check if kube gateway integration is enabled by checking if the Gateway API CRDs are installed and
+// whether a GatewayParameters CR exists in the install namespace
+func isKubeGatewayEnabled(ctx context.Context, opts *options.Options) (bool, error) {
```
Comment (davidjumani) [inline projects/gloo/cli/pkg/cmd/check/kube_gateway.go:55]: If the user decides to test the product and initially installs it in the kube gateway mode, the k8s gateway CRDs and the GatewayParameters will be installed. 
Later on, if they decide to change the just `kubeGateway.enabled` helm value (and not do a clean uninstall) to install it in the regular mode, the k8s gateway CRDs and the GatewayParameters would still exist.
Would this check be able to handle this scenario ? If not, could we rely on something such as the presence of the `GG_EXPERIMENTAL_K8S_GW_CONTROLLER` env var in the gloo deployment or something that is guaranteed to change between installs if the helm setting is toggled ?

**Comment by davidjumani** [comment]
I have a similar approach and would like feedback on it :
- First, check the presence of the GatewayParameters CR.
- If found, check the presence of the k8s gateway CRDs.
- If they are not found, error loudly that they are not found and the installation is misconfigured as the GatewayParameters CR is present but the k8s gateway CRDs are not.
- Else proceed with the k8s gateway-related checks

Additionally I think that relying on the presence of the GatewayParameters CR wouldn't be the best way to go about this as users can delete the default CRs and replace them with their own.
Instead, we can label the gloo deployment and check for the presence of that label or the presence of the GG_EXPERIMENTAL_K8S_GW_CONTROLLER env variable something more concrete. I understand that this is tying up the check with the implementation but it seems more robust

File: test/kubernetes/e2e/tests/k8s_gw_tests.go [hunk 10]
```diff
@@ -51,7 +51,17 @@ func newGlooctlTestingSuite(ctx context.Context, testInstallation *e2e.TestInsta
...
+	// 0. Install Edge with kubeGateway enabled (done when the test begins)
+	// 1. Verify it checks kubeGateway resources
 	suite.Run(s.T(), glooctl.NewCheckSuite(s.ctx, s.testInstallation))
+	// TODO (davidjumani) :
+	// 2. Upgrade Edge with kubeGateway disabled
+	// 3. Verify it does not check kubeGateway resources
+	// 4. Upgrade Edge with kubeGateway enabled
+	// 5. Verify it checks kubeGateway resources
+	// This verifies that we are not relying on any logic / resources that can be left behind after an upgrade or when the user switches between gateway modes
+	// Doing so will also eliminate the need for the kube2e/glooctl/exclude tests
```
Comment (davidjumani) [inline test/kubernetes/e2e/tests/k8s_gw_tests.go:64]: Manually verified this works. Would be nice to add a test in a follow up

File: projects/gloo/cli/pkg/cmd/check/internal/checks.go [hunk 11]
```diff
@@ -6,120 +6,25 @@ import (
...
-		printer.AppendStatus("deployments", fmt.Sprintf("%v Errors!", multiErr.Len()))
-		return multiErr.ErrorOrNil()
-	}
-	printer.AppendStatus("deployments", "OK")
-	return nil
-}
-
 func CheckGatewayClass(ctx context.Context, printer printers.P, opts *options.Options) error {
-	printer.AppendCheck("Checking GatewayClass... ")
+	printer.AppendCheck("Checking kube GatewayClasses... ")
```
Comment (npolshakova) [inline projects/gloo/cli/pkg/cmd/check/internal/checks.go:21]: [nit] Do we want to name this `Kubernetes GatewayClasses`, `K8s GatewayClasses` or `K8s GatewayAPI GatewayClasses`?

File: projects/gloo/cli/pkg/cmd/check/kube_gateway.go [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+)
+
+type CheckFunc = func(ctx context.Context, printer printers.P, opts *options.Options) error
+
+func CheckKubeGatewayResources(ctx context.Context, printer printers.P, opts *options.Options) error {
+	var multiErr *multierror.Error
+
+	kubeGatewayEnabled, err := isKubeGatewayEnabled(ctx, opts)
+	if err != nil {
+		multiErr = multierror.Append(multiErr, eris.Wrapf(err, "unable to determine if kube gateway is enabled"))
```
Comment (npolshakova) [inline projects/gloo/cli/pkg/cmd/check/kube_gateway.go:23]: [nit] Maybe have consistent naming? 
- `Skipping Kubernetes Gateway resources check -- Kubernetes Gateway integration not enabled"`
- `unable to determine if Kubernetes Gateway integration is enabled`

File: projects/gloo/cli/pkg/cmd/check/kube_gateway.go [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+// whether the GG_EXPERIMENTAL_K8S_GW_CONTROLLER env var is true in the gloo deployment.
+func isKubeGatewayEnabled(ctx context.Context, opts *options.Options) (bool, error) {
+	cfg, err := kubeutils.GetRestConfigWithKubeContext(opts.Top.KubeContext)
+	if err != nil {
+		return false, err
+	}
+
+	gatewayEnabled, err := kubegatewayutils.DetectKubeGatewayEnabled(ctx, opts)
+	if err != nil {
+		return false, eris.Wrapf(err, "could not determine if kubeGateway is enabled")
```
Comment (npolshakova) [inline projects/gloo/cli/pkg/cmd/check/kube_gateway.go:67]: [nit] Consistent naming? 
```suggestion
		return false, eris.Wrapf(err, "could not determine if Kubernetes Gateway integration is enabled")
```

File: projects/gloo/cli/pkg/cmd/check/kube_gateway.go [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+	}
+
+	if !gatewayEnabled {
+		return false, nil
+	}
+
+	hasCRDs, err := kubegatewayutils.DetectKubeGatewayCrds(cfg)
+	if err != nil {
+		return false, eris.Wrapf(err, "could not determine if kubernetes gateway crds are applied")
```
Comment (npolshakova) [inline projects/gloo/cli/pkg/cmd/check/kube_gateway.go:76]:
```suggestion
		return false, eris.Wrapf(err, "could not determine if Kubernetes Gateway CRDs are applied")
```

File: projects/gloo/cli/pkg/cmd/check/kube_gateway.go [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+	}
+	if !hasCRDs {
+		return false, eris.New("KubeGateway is enabled but the Kubernetes Gateway CRDs are not applied. Please fix this by running `kubectl apply -f https://github.com/kubernetes-sigs/gateway-api/releases/download/v1.0.0/standard-install.yaml`")
```
Comment (npolshakova) [inline projects/gloo/cli/pkg/cmd/check/kube_gateway.go:79]:
```suggestion
		return false, eris.New("Kubernetes Gateway integration is enabled but the Kubernetes Gateway CRDs are not applied. Please fix this by running `kubectl apply -f https://github.com/kubernetes-sigs/gateway-api/releases/download/v1.0.0/standard-install.yaml`")
```

Comment (npolshakova) [reply projects/gloo/cli/pkg/cmd/check/kube_gateway.go:79]: Do we want to keep the link in the error message? This could change over time?

File: projects/gloo/cli/pkg/kubegatewayutils/detect.go [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+	"github.com/solo-io/gloo/projects/gloo/cli/pkg/helpers"
+	"github.com/solo-io/gloo/projects/gloo/constants"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/rest"
+)
+
+const (
+	KubeGatewayNetworkingGroup = "gateway.networking.k8s.io"
```
Comment (npolshakova) [inline projects/gloo/cli/pkg/kubegatewayutils/detect.go:20]: Why not reuse? https://github.com/solo-io/gloo/blob/40611c38cac8cd6cd2dba8088cb4082b3f265bea/projects/gateway2/wellknown/gwapi.go#L5

File: test/kubernetes/e2e/tests/k8s_gw_tests.go (hunk 10 above)
Comment (npolshakova) [reply test/kubernetes/e2e/tests/k8s_gw_tests.go:64]: Yep, adding a Gloo Edge test doesn't check K8s Gateway resources would be a good follow up! We already have a Gloo Edge helm manifest, we would just need to add the inverse checks that the kubeGateway resources do not get checked. https://github.com/solo-io/gloo/blob/406710a7cdd4cd1ace24d53247f4c937cf8d3589/test/kubernetes/e2e/tests/edge_gw_test.go

I don't know if it's worth having an e2e test for an invalid installation. We can just manually validate it and make sure there are unit tests for coverage.

File: projects/gloo/cli/pkg/cmd/check/internal/checks.go (hunk 11 above)
Comment (davidjumani) [reply projects/gloo/cli/pkg/cmd/check/internal/checks.go:21]: I'll rename it to `Kubernetes GatewayClasses`

File: projects/gloo/cli/pkg/cmd/check/kube_gateway.go (hunk 12 above)
Comment (davidjumani) [reply projects/gloo/cli/pkg/cmd/check/kube_gateway.go:23]: Updated

File: projects/gloo/cli/pkg/cmd/check/kube_gateway.go (hunk 13 above)
Comment (davidjumani) [reply projects/gloo/cli/pkg/cmd/check/kube_gateway.go:67]: updated

File: projects/gloo/cli/pkg/cmd/check/kube_gateway.go (hunk 14 above)
Comment (davidjumani) [reply projects/gloo/cli/pkg/cmd/check/kube_gateway.go:76]: updated

File: projects/gloo/cli/pkg/cmd/check/kube_gateway.go (hunk 15 above)
Comment (davidjumani) [reply projects/gloo/cli/pkg/cmd/check/kube_gateway.go:79]: Updated. I think we can keep it here for now. I prefer setting the link to `https://github.com/kubernetes-sigs/gateway-api/releases/latest/download/standard-install.yaml` so there's no need to update it but unsure up to which version we support right now

File: projects/gloo/cli/pkg/kubegatewayutils/detect.go (hunk 16 above)
Comment (davidjumani) [reply projects/gloo/cli/pkg/kubegatewayutils/detect.go:20]: Updated

File: test/kubernetes/e2e/tests/k8s_gw_tests.go (hunk 10 above)
Comment (davidjumani) [reply test/kubernetes/e2e/tests/k8s_gw_tests.go:64]: I've added the inverse tests as part of the glooctl kube2e suie [here](https://github.com/solo-io/gloo/pull/9621/files#diff-2775a6b97e3b5beedb2758cf5a2a162088ad9870643f5bea7e605345620d5070R196-R213) but doing it during upgrades switching between gateway modes will better validate the behaviour

File: projects/gloo/cli/pkg/cmd/check/kube_gateway.go [hunk 17]
```diff
@@ -0,0 +1,83 @@
...
+	var multiErr *multierror.Error
+
+	kubeGatewayEnabled, err := isKubeGatewayEnabled(ctx, opts)
+	if err != nil {
+		multiErr = multierror.Append(multiErr, eris.Wrapf(err, "unable to determine if Kubernetes Gateway integration is enabled"))
+		return multiErr
+	}
+
+	if !kubeGatewayEnabled {
+		printer.AppendMessage("\nSkipping Kubernetes Gateway resources check -- Kube Gateway integration not enabled")
```
Comment (npolshakova) [inline projects/gloo/cli/pkg/cmd/check/kube_gateway.go:28]: I think this one still has `Kube Gateway`

File: test/kubernetes/e2e/tests/k8s_gw_tests.go (hunk 10 above)
Comment (npolshakova) [reply test/kubernetes/e2e/tests/k8s_gw_tests.go:64]: I think the link isn't working? Are these the tests in test/kubernetes/e2e/features/glooctl/check.go? I meant adding a new suite here for glooctl without kube gateways: https://github.com/solo-io/gloo/blob/406710a7cdd4cd1ace24d53247f4c937cf8d3589/test/kubernetes/e2e/tests/edge_gw_tests.go 

We currently don't have an upgrade test in the new testing framework, so building that out would be a bigger issue.

File: projects/gloo/cli/pkg/cmd/check/kube_gateway.go (hunk 15 above)
Comment (npolshakova) [reply projects/gloo/cli/pkg/cmd/check/kube_gateway.go]: Yeah, I wonder if linking to the installation doc on our docs would be better? This isn't a blocker but might be good to get docs input- @Nadine2016 do you have any thoughts?

File: test/kubernetes/e2e/tests/k8s_gw_tests.go (hunk 10 above)
Comment (davidjumani) [reply test/kubernetes/e2e/tests/k8s_gw_tests.go:64]: They are in `test/kube2e/glooctl/check_test.go` that runs glooctl on kubeGateway=disabled installs. It verifies that it does not check the Kubernetes Gateway resources

File: projects/gloo/cli/pkg/cmd/check/kube_gateway.go (hunk 15 above)
Comment (Nadine2016) [reply projects/gloo/cli/pkg/cmd/check/kube_gateway.go]: Yeah I think in this particular case it would be easier for the user to have the exact command handy. If we feel however that this can change often, then I would point them to our docs instead. 

If we want to keep it, then we could change the text to
The Kubernetes Gateway integration is enabled, but the Kubernetes Gateway CRDs are not applied in the cluster. To apply the CRDs, run `kubectl apply -f https://github.com/kubernetes-sigs/gateway-api/releases/download/v1.0.0/standard-install.yaml`.

Comment (davidjumani) [reply projects/gloo/cli/pkg/cmd/check/kube_gateway.go]: Updated

File: projects/gloo/cli/pkg/cmd/check/kube_gateway.go [hunk 18]
```diff
@@ -0,0 +1,83 @@
...
+	if !gatewayEnabled {
+		return false, nil
+	}
+
+	hasCRDs, err := kubegatewayutils.DetectKubeGatewayCrds(cfg)
+	if err != nil {
+		return false, eris.Wrapf(err, "unable to determine if Kubernetes Gateway CRDs are applied")
+	}
+	if !hasCRDs {
+		return false, eris.New("The Kubernetes Gateway integration is enabled, but the Kubernetes Gateway CRDs are not applied in the cluster. To apply the CRDs, run `kubectl apply -f https://github.com/kubernetes-sigs/gateway-api/releases/download/v1.0.0/standard-install.yaml`")
```
Comment (nfuden) [inline projects/gloo/cli/pkg/cmd/check/kube_gateway.go:79]: Can we link this to somewhere in our code base with the version that we test against?

**Review by sam-heilbron — COMMENTED** [review]
We have some new CRDs:
- https://github.com/solo-io/gloo/blob/main/install/helm/gloo/crds/gateway.solo.io_v1_HttpListenerOption.yaml
- https://github.com/solo-io/gloo/blob/main/install/helm/gloo/crds/gateway.solo.io_v1_ListenerOption.yaml
They don't report statuses, but I think it would make sense to check them still

**Review by sam-heilbron — APPROVED** [review]

File: projects/gloo/cli/pkg/cmd/check/kube_gateway.go (hunk 18 above)
Comment (davidjumani) [reply projects/gloo/cli/pkg/cmd/check/kube_gateway.go:79]: Will do in a followup

**Comment by davidjumani** [comment]
kick bulldozer

**Comment by davidjumani** [comment]
> We have some new CRDs:
> 
>     * https://github.com/solo-io/gloo/blob/main/install/helm/gloo/crds/gateway.solo.io_v1_HttpListenerOption.yaml
> 
>     * https://github.com/solo-io/gloo/blob/main/install/helm/gloo/crds/gateway.solo.io_v1_ListenerOption.yaml
>       They don't report statuses, but I think it would make sense to check them still

Will tackle in a followup PR
